### DISCUSSION TITLE
marksman lsp erroring with perm denied trying to read some homedir directories

### DISCUSSION THREAD

**Discussion by bytebeast** [artempyanykh/marksman#377] [General]
these errors seem to only occur when i open a *.md file in my parent home directory. unclear why its trying to read all the sub-directories.

sample entries from lsp.log:

```
[ERROR][2024-12-11 20:22:36] .../vim/lsp/rpc.lua:759	"rpc"	"/Users/<MY_USERNAME>/.local/share/nvim/mason/bin/marksman"	"stderr"      	"[20:22:36 WRN] <Folder> Couldn't read the folder: {\"dir\": \"Abs\\n  (AbsPath\\n     \\\"/Users/<MY_USERNAME>/Library/Containers/com.apple.icloud.apps.messages.business.extension/Data/Music/Music\\\")\"}\nSystem.UnauthorizedAccessException: Access to the path '/Users/<MY_USERNAME>/Library/Containers/com.apple.icloud.apps.messages.business.extension/Data/Music/Music' is denied.\n ---> System.IO.IOException: Operation not permitted\n   --- End of inner exception stack trace ---\n   at System.IO.Enumeration.FileSystemEnumerator`1.Init()\n   at System.IO.Enumeration.FileSystemEnumerableFactory.FileInfos(String, String, EnumerationOptions, Boolean)\n   at System.IO.DirectoryInfo.InternalEnumerateInfos(String, String, SearchTarget, EnumerationOptions)\n   at Marksman.Folder.FolderModule.collect@303(String[] configuredExts, UriWith`1 folderId, LocalPath cur, FSharpList`1 ignoreMatchers) in /Users/runner/work/marksman/marksman/Marksman/Folder.fs:line 312\n"
[ERROR][2024-12-11 20:22:36] .../vim/lsp/rpc.lua:759	"rpc"	"/Users/<MY_USERNAME>/.local/share/nvim/mason/bin/marksman"	"stderr"      	"[20:22:36 WRN] <Folder> Couldn't read the folder: {\"dir\": \"Abs\\n  (AbsPath\\n     \\\"/Users/<MY_USERNAME>/Library/Containers/com.apple.icloud.apps.messages.business.extension/Data/Library/Calendars\\\")\"}\nSystem.UnauthorizedAccessException: Access to the path '/Users/<MY_USERNAME>/Library/Containers/com.apple.icloud.apps.messages.business.extension/Data/Library/Calendars' is denied.\n ---> System.IO.IOException: Operation not permitted\n   --- End of inner exception stack trace ---\n   at System.IO.Enumeration.FileSystemEnumerator`1.Init()\n   at System.IO.Enumeration.FileSystemEnumerableFactory.FileInfos(String, String, EnumerationOptions, Boolean)\n   at System.IO.DirectoryInfo.InternalEnumerateInfos(String, String, SearchTarget, EnumerationOptions)\n   at Marksman.Folder.FolderModule.collect@303(String[] configuredExts, UriWith`1 folderId, LocalPath cur, FSharpList`1 ignoreMatchers) in /Users/runner/work/marksman/marksman/Marksman/Folder.fs:line 312\n"
[ERROR][2024-12-11 20:22:36] .../vim/lsp/rpc.lua:759	"rpc"	"/Users/<MY_USERNAME>/.local/share/nvim/mason/bin/marksman"	"stderr"      	"[20:22:36 WRN] <Folder> Couldn't read the folder: {\"dir\": \"Abs\\n  (AbsPath\\n     \\\"/Users/<MY_USERNAME>/Library/Containers/com.apple.icloud.apps.messages.business.extension/Data/Library/Mail\\\")\"}\nSystem.UnauthorizedAccessException: Access to the path '/Users/<MY_USERNAME>/Library/Containers/com.apple.icloud.apps.messages.business.extension/Data/Library/Mail' is denied.\n ---> System.IO.IOException: Operation not permitted\n   --- End of inner exception stack trace ---\n   at System.IO.Enumeration.FileSystemEnumerator`1.Init()\n   at System.IO.Enumeration.FileSystemEnumerableFactory.FileInfos(String, String, EnumerationOptions, Boolean)\n   at System.IO.DirectoryInfo.InternalEnumerateInfos(String, String, SearchTarget, EnumerationOptions)\n   at Marksman.Folder.FolderModule.collect@303(String[] configuredExts, UriWith`1 folderId, LocalPath cur, FSharpList`1 ignoreMatchers) in /Users/runner/work/marksman/marksman/Marksman/Folder.fs:line 312\n"
[ERROR][2024-12-11 20:22:36] .../vim/lsp/rpc.lua:759	"rpc"	"/Users/<MY_USERNAME>/.local/share/nvim/mason/bin/marksman"	"stderr"
```

I don't have marksman manually configured in `~/.config/nvim/lua/plugins`. As far as I can tell it gets loaded from  `~/.local/share/nvim/lazy/LazyVim/lua/lazyvim/plugins/extras/lang/markdown.lua`

Not sure if this a bug, or if its something i unknowingly broke on my setup.
```
~/.local/share/nvim/mason/bin/marksman --version
1.0.0-ea06e98+ea06e9813e71a7611b78f7781b7a70a901a00916
```

I don't have any lsp configs in plugins. Im currently only using whats loaded through extras.

These errors only appear when editing a markdown file in my home directory which leads me to believe it recursively scans all subdirectories of the file your editing. Guessing this is becasue you can link to other markdown files from a single markdown file?

Is there a way to disable the recursive scans from specific directories OR only enable recursive scans of specific directories OR maybe just the ability to disable recursive scans all together?

**Comment by artempyanykh** [comment]
I have a feeling that it's related to #375. At least, that's not the first time I see nvim with mason behaving weirdly. @bytebeast could you show the part of the logs that corresponds to the LSP initialization request? The relevant entry has `method = "initialize"` and then `rootPath` and `rootUri`.

> Is there a way to disable the recursive scans from specific directories OR only enable recursive scans of specific directories OR maybe just the ability to disable recursive scans all together?

Let me give a bit more color on this. Marksman will try to do a recursive scan only within a workspace folder that is communicated by the LSP client (an editor). More specifically, when neovim start marksman, it sends it an initialization request. As part of this request's payload it sends rootUri or rootPath. Then:
* If the root is null, Marksman will enable single-file mode: no recursive scans, LSP features work only in the context of a single opened file.
* If it's not null, then Marksman will also check that the root is not bogus (because different editors behave differently): it'll look either for `.marksman.toml` or `.git` right under the root folder. If found, then it'll be enable multi-file mode and do a recursive scan for other documents. If not found, it'll be single-file mode.

To experience the behavior as you described, nvim would need to send your home dir as `rootUri` during marksman's initialization **and** there should be either `.marksman.toml` or `.git` folders under your home dir. And this is why I'd need to see the initialization request.

More on this in: https://github.com/artempyanykh/marksman/blob/main/docs/features.md#workspace-folders-project-roots-and-single-file-mode

**Comment by artempyanykh** [comment]
@bytebeast perhaps easier, the output of `:LspInfo` contains the root too.

**Comment by artempyanykh** [comment]
I just noticed that this is LazyVim. And I seem to be able to reproduce the problem with stock LazyVim on my end.

**Comment by artempyanykh** [comment]
Ok, I found that "home folder as a root" comes from the nvim sending the following notification:
```
[DEBUG][2024-12-13 21:13:56] .../vim/lsp/rpc.lua:286	"rpc.send"	{  jsonrpc = "2.0",  method = "workspace/didChangeWorkspaceFolders",  params = {    event = {      added = { {          name = "/Users/<USERNAME>",          uri = "file:///Users/<USERNAME>"        } },      removed = {}    }  }}
```

I don't have this with my own config. Unclear, where does this come from.

**Comment by bytebeast** [comment]
its just the current working directory where the foo.md file is opened.
guessing it gets passed in via lazyvim(mason) ?  ( /shrug )

errors only seen when opening md files in my home directory. i have a
vanilla lazyvim config with markdown "extra" plugin provided by lazyvim.

**Comment by artempyanykh** [comment]
I think #380 should address it. Also https://github.com/LazyVim/LazyVim/issues/5074
